### PR TITLE
Treeshake instanceof

### PR DESCRIPTION
**Pull request by lukastaegert**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### Description
This is an approach to evaluate "instanceof" as "false" if the right hand side is not included, see here: https://rollupjs.org/repl/?pr=4524&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNsYXNzJTIwVW51c2VkJTIwJTdCJTdEJTVDbmNsYXNzJTIwVXNlZCUyMCU3QiU3RCU1Q24lNUNuaWYlMjAobmV3JTIwVXNlZCgpJTIwaW5zdGFuY2VvZiUyMFVudXNlZCklMjBjb25zb2xlLmxvZygncmVtb3ZlZCcpJTNCJTVDbmVsc2UlMjBjb25zb2xlLmxvZygncmV0YWluZWQnKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

### CODE REVIEW

File: test/form/samples/instanceof/_expected.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+console.log('retained');
+
+console.log('retained');
```
Comment (LeviPesin) [inline test/form/samples/instanceof/_expected.js:3]: Shouldn't there also be `console.log('effect')`, because the effect should be retained?

Comment (lukastaegert) [reply test/form/samples/instanceof/_expected.js:3]: You are right, not sure how I missed this this morning. In any case, this is more of a draft that needs more polishing in other places as well.

Comment (lukastaegert) [reply test/form/samples/instanceof/_expected.js:3]: Everything should be working as expected now

**Comment by lukastaegert** [comment]
There is one caveat here in that for technical reasons, we need to include all variables referenced in the left-hand side of `instanceof` because otherwise I do not know how to catch cases where the only usage of a class is inside the `instanceof` left-hand side, i.e.

```js
class Foo {}
if (new Foo() instanceof Foo) console.log('must be retained');
```

**Comment by lukastaegert** [comment]
Ok, I take back my previous assessment. The problem is that the slightly hacky "include" logic does not work through parameters, which is a really important core use case, see [this example](https://rollupjs.org/repl/?pr=4524&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmxldCUyMGVmZmVjdCUyMCUzRCUyMGZhbHNlJTNCJTVDbiU1Q25jbGFzcyUyMEZvbyUyMCU3QiU3RCU1Q24lNUNuZnVuY3Rpb24lMjBjaGVja0luc3RhbmNlKGluc3RhbmNlKSUyMCU3QiU1Q24lNUN0aWYlMjAoaW5zdGFuY2UlMjBpbnN0YW5jZW9mJTIwRm9vKSUyMCU3QiU1Q24lNUN0JTVDdGVmZmVjdCUyMCUzRCUyMHRydWUlM0IlNUNuJTVDdCU3RCU1Q24lN0QlNUNuJTVDbmNoZWNrSW5zdGFuY2UobmV3JTIwRm9vKCkpJTNCJTVDbmFzc2VydC5vayhlZmZlY3QpJTNCY2xhc3MlMjBVbnVzZWQlMjAlN0IlN0QlNUNuJTVDbmNsYXNzJTIwVXNlZCUyMCU3QiU3RCU1Q25pZiUyMChuZXclMjBVc2VkKCklMjBpbnN0YW5jZW9mJTIwVW51c2VkKSUyMCU3QiU3RCUzQiU1Q24lNUNuZnVuY3Rpb24lMjBjaGVja0luc3RhbmNlKGluc3RhbmNlKSUyMCU3QiU1Q24lMjAlMjBpZiUyMChpbnN0YW5jZSUyMGluc3RhbmNlb2YlMjBVbnVzZWQpJTIwJTdCJTVDbiU1Q3QlNUN0Y29uc29sZS5sb2coJ3JlbW92ZWQnKSUzQiU1Q24lNUN0JTdEJTNCJTVDbiU3RCU1Q25jaGVja0luc3RhbmNlKCU3QiU3RCklM0IlNUNuJTVDbmNvbnN0JTIwb2JqJTIwJTNEJTIwJTdCJTdEJTNCJTVDbmlmJTIwKG9iaiUyMGluc3RhbmNlb2YlMjBVbnVzZWQpJTIwJTdCJTdEJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==) where the effect is omitted.
While it might be possible to fix, this feels very fragile to me and I will hold off this solution for now until I have a fundamentally better idea. Still, I will keep the regression tests I created.

**Pull request closed without merging**